src/search/tasks/root_task.cc

```diff
@@ -179,84 +178,104 @@ ExplicitEffect::ExplicitEffect(
     if (value_pre != -1) {
         preconditions.emplace_back(var, value_pre);
     }
     effects.emplace_back(var, value_post, move(conditions));
 }
 
-ExplicitOperator::ExplicitOperator(istream &in, bool is_an_axiom, bool use_metric)
+ExplicitOperator::ExplicitOperator(utils::InputFileParser &in, bool is_an_axiom, bool use_metric)
     : is_an_axiom(is_an_axiom) {
     if (!is_an_axiom) {
-        check_magic(in, "begin_operator");
-        in >> ws;
-        getline(in, name);
-        preconditions = read_facts(in);
-        int count;
-        in >> count;
+        in.read_magic_line("begin_operator");
+        name = in.read_line("operator name");
+        preconditions = read_facts(in, false);
+        int count = in.read_line_int("number of operator effects");
         effects.reserve(count);
         for (int i = 0; i < count; ++i) {
             read_pre_post(in);
         }
-
-        int op_cost;
-        in >> op_cost;
+        int op_cost = in.read_line_int("operator cost");
         cost = use_metric ? op_cost : 1;
-        check_magic(in, "end_operator");
+        in.read_magic_line("end_operator");
     } else {
         name = "<axiom>";
         cost = 0;
-        check_magic(in, "begin_rule");
-        read_pre_post(in);
-        check_magic(in, "end_rule");
+        in.read_magic_line("begin_rule");
+        read_axiom(in);
+        in.read_magic_line("end_rule");
     }
     assert(cost >= 0);
 }
 
-static void read_and_verify_version(istream &in) {
-    int version;
-    check_magic(in, "begin_version");
-    in >> version;
-    check_magic(in, "end_version");
+static void read_and_verify_version(utils::InputFileParser &in) {
+    in.set_context("version section");
+    in.read_magic_line("begin_version");
+    int version = in.read_line_int("version number");
     if (version != PRE_FILE_VERSION) {
         cerr << "Expected translator output file version " << PRE_FILE_VERSION
              << ", got " << version << "." << endl
              << "Exiting." << endl;
         utils::exit_with(ExitCode::SEARCH_INPUT_ERROR);
     }
-}
-
-static bool read_metric(istream &in) {
-    bool use_metric;
-    check_magic(in, "begin_metric");
-    in >> use_metric;
-    check_magic(in, "end_metric");
+    in.read_magic_line("end_version");
+}
+
+static bool read_metric(utils::InputFileParser &in) {
+    in.set_context("metric_section");
+    in.read_magic_line("begin_metric");
+    string use_metric_string = in.read_line("use metric");
+    bool use_metric = false;
+    if (use_metric_string == "1") {
+        use_metric = true;
+    } else if (use_metric_string == "0") {
+        use_metric = false;
+    } else {
+        in.error("expected boolean");
+    }
+    in.read_magic_line("end_metric");
     return use_metric;
 }
 
-static vector<ExplicitVariable> read_variables(istream &in) {
-    int count;
-    in >> count;
+static vector<ExplicitVariable> read_variables(utils::InputFileParser &file_parser) {
```

Comment:
The parser is sometimes called `in` and sometimes `file_parser`. I'm fine with both but would prefer it to be consistent.